File: projects/spring-boot-restv2/demo/src/test/java/com/sourabh/demo/servicetest/CustomerServiceTest.java
package com.sourabh.demo.servicetest;

import com.sourabh.demo.exceptions.SomethingNotFoundException;
import com.sourabh.demo.model.Customer;
import com.sourabh.demo.model.Room;
import com.sourabh.demo.model.Transaction;
import com.sourabh.demo.repository.CustomerRepository;
import com.sourabh.demo.service.CustomerService;
import com.sourabh.demo.service.RoomService;
import com.sourabh.demo.service.SequenceGeneratorService;
import com.sourabh.demo.service.TransactionService;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.mock.mockito.MockBean;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import static org.mockito.Mockito.when;

@SpringBootTest
class CustomerServiceTest {

    @Autowired
    CustomerService customerService;

    @MockBean
    RoomService roomService;
    @MockBean
    CustomerRepository customerRepository;
    @MockBean
    TransactionService transactionService;
    @MockBean
    SequenceGeneratorService sequenceGenerator;
    @MockBean
    Transaction transaction;

    @Test
    void addCustomerTest(){
        Customer customer = new Customer();
        customer.setName("Sourabh");
        customer.setType("Suit");
        customer.setDays(5);
        customer.setPaymentMethod("UPI");
        customer.setFullPay(true);
        Room room = new Room(1,"Suit",1200,12,true);
        when(roomService.findByTypeAvailable("Suit")).thenReturn(room);
        when(sequenceGenerator.generateSequence(Customer.SEQUENCE_NAME)).thenReturn(1);
        Customer modifiedCustomer = new Customer();
        modifiedCustomer.setName("Sourabh");
        modifiedCustomer.setType("Suit");
        modifiedCustomer.setDays(5);
        modifiedCustomer.setPaymentMethod("UPI");
        modifiedCustomer.setFullPay(true);
        modifiedCustomer.setCid(101);
        modifiedCustomer.setTotalFair(4800);
        modifiedCustomer.setStatus(true);
        modifiedCustomer.setBookDate(LocalDateTime.now());
        when(customerRepository.save(customer)).thenReturn(modifiedCustomer);
        Assertions.assertEquals(modifiedCustomer,customerService.addCustomer(customer));
    }

    @Test
    void addCustomerTest2(){
        Customer customer = new Customer();
        customer.setName("Sourabh");
        customer.setType("Suit");
        customer.setDays(0);
        customer.setPaymentMethod("UPI");
        customer.setFullPay(true);
        Assertions.assertThrows(SomethingNotFoundException.class, ()->customerService.addCustomer(customer));
    }

    @Test
    void getAllTest(){

        List<Customer> customers = new ArrayList<>();
        Customer customer1 = new Customer(1,"OtherMe","Suit",6,6000,true,"UPI", LocalDateTime.now(),true);
        Customer customer2 = new Customer(2,"SourabhOther","Suit",6,6000,true,"UPI", LocalDateTime.now(),true);
        Customer customer3 = new Customer(3,"HelloMe","Suit",6,6000,true,"UPI", LocalDateTime.now(),true);
        Customer customer4 = new Customer(4,"Sourabh4","Single-Room",6,6000,true,"UPI", LocalDateTime.now(),true);
        customers.add(customer1);
        customers.add(customer2);
        customers.add(customer3);
        customers.add(customer4);
        when(customerRepository.findAll()).thenReturn(customers);
        Assertions.assertEquals(4,customerService.getAll().size());
    }
    @Test
    void getCustomerTest(){
        Customer customer = new Customer(1,"Sourabh","Suit",6,6000,true,"UPI", LocalDateTime.now(),true);
        when(customerRepository.findById(1)).thenReturn(java.util.Optional.of(customer));
        Assertions.assertEquals(customer.getName(),customerService.getCustomer(1).getName());

    }

    @Test
    void calculateFareTest(){
        int days = 5;
        float baseFare = 1200.0F;
        float expected = 4800.0F;
        Assertions.assertEquals(expected,customerService.calculateFare(days,baseFare));

    }







}

File: projects/spring-boot-rest-kt/hotel-reservation/src/test/kotlin/com/sourabh/hotelreservation/servicetest/CustomerServiceTest.kt
package com.sourabh.hotelreservation.servicetest

import com.sourabh.hotelreservation.exception.SomethingNotFoundException
import com.sourabh.hotelreservation.model.Customer
import com.sourabh.hotelreservation.model.Room
import com.sourabh.hotelreservation.repository.CustomerRepository
import com.sourabh.hotelreservation.service.CustomerService
import com.sourabh.hotelreservation.service.RoomService
import com.sourabh.hotelreservation.service.SequenceGeneratorService
import com.sourabh.hotelreservation.service.TransactionService
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.springframework.boot.test.context.SpringBootTest
import java.time.LocalDateTime
import java.util.*
import kotlin.collections.ArrayList

@SpringBootTest
class CustomerServiceTest {
    @InjectMocks
    var customerService: CustomerService? = null

    @Mock
    var roomService: RoomService? = null

    @Mock
    var customerRepository: CustomerRepository? = null

    @Mock
    var transactionService: TransactionService? = null

    @Mock
    var sequenceGenerator: SequenceGeneratorService? = null

    @Test
    fun addCustomerTest() {
        val customer = Customer(name = "Sourabh", type = "Suit", fullPay = true, days = 5, paymentMethod = "UPI")
        val room = Room(1, "Suit", 1200F, 12, true)
        Mockito.`when`(roomService!!.findByTypeAvailable("Suit")).thenReturn(room)
        Mockito.`when`(sequenceGenerator!!.generateSequence(Customer.SEQUENCE_NAME)).thenReturn(1)
        val modifiedCustomer = Customer()
        modifiedCustomer.name="Sourabh"
        modifiedCustomer.type="Suit"
        modifiedCustomer.days=5
        modifiedCustomer.paymentMethod="UPI"
        modifiedCustomer.status=true
        modifiedCustomer.cid=101
        modifiedCustomer.totalFair=4800F
        modifiedCustomer.fullPay=true
        modifiedCustomer.bookDate=LocalDateTime.now()
        Mockito.`when`(customerRepository!!.save(customer)).thenReturn(modifiedCustomer)
        Assertions.assertEquals(modifiedCustomer, customerService!!.addCustomer(customer))
    }

    @Test
    fun addCustomerTest2() {
        val customer = Customer(name = "Sourabh", type = "Suit", fullPay = true, days = 0, paymentMethod = "UPI")
        Assertions.assertThrows(SomethingNotFoundException::class.java) { customerService?.addCustomer(customer) }
    }

        @Test
    fun getAllTest() {
        val customers: MutableList<Customer> = ArrayList()
        val customer1 = Customer(1, "Sourabh", "Suit", 6, 6000F, true, LocalDateTime.now(), "UPI", true)
        val customer2= Customer(2, "SourabhOther", "Single-Bed", 6, 6000F, true, LocalDateTime.now(), "UPI", true)
        val customer3= Customer(3, "Hello-again", "Dormitory", 5, 6000F, true, LocalDateTime.now(), "UPI", true)
        customers.add(customer1)
        customers.add(customer2)
        customers.add(customer3)
        Mockito.`when`(customerRepository!!.findAll()).thenReturn(customers)
        Assertions.assertEquals(3, customerService?.getAll()?.size)
    }
    @Test
    fun getCustomerTest(){
        val customer=Customer(1, "Sourabh", "Suit", 6, 6000F, true, LocalDateTime.now(), "UPI", true)
        Mockito.`when`(customerRepository!!.findById(1)).thenReturn(Optional.of(customer))
        Assertions.assertEquals(customer.name, customerService!!.getCustomer(1).name)
    }
    @Test
    fun calculateFareTest(){
        val days=5;
        val baseFare = 1200F
        val expected = 4800F
        Assertions.assertEquals(expected,customerService!!.calculateFare(days,baseFare))
    }


}